mrp/src/mrp/runtime/conda.py

```diff
@@ -278,64 +284,72 @@ def _env_name(self):
 
-        def _cache_valid(self, yaml_path, env_content):
-            """Check if the cache is valid.
+        def _config_path(self):
+            dirpath = os.path.expanduser(f"~/.config/mrp/conda/{self._env_name()}/")
+            os.makedirs(dirpath, exist_ok=True)
+            return dirpath
 
-            To be valid, the cache must:
-            - have the named yaml file.
-            - have the same channels and dependencies as generated in the last up command.
-            - have a single command in the conda env history.
-            - the command must match the creation command.
-            """
-            try:
-                if not os.path.exists(yaml_path):
-                    return False
+        def _yaml_path(self):
+            return os.path.join(self._config_path(), "conda_env.yaml")
 
-                old_content = json.load(open(yaml_path))
+        def _setup_commands_snapshot_path(self):
+            return os.path.join(self._config_path(), "setup_commands.snapshot")
 
-                old_content_compare_key = json.dumps(old_content, sort_keys=True)
-                new_content_compare_key = json.dumps(env_content, sort_keys=True)
-                if old_content_compare_key != new_content_compare_key:
-                    return False
+        def _conda_history_snapshot_path(self):
+            return os.path.join(self._config_path(), "history.snapshot")
 
-                info = json.loads(
-                    subprocess.check_output(
-                        ["conda", "env", "export", "--json", "-n", self._env_name()]
-                    )
+        def _conda_history_path(self):
+            info = json.loads(
+                subprocess.check_output(
+                    ["conda", "env", "export", "--json", "-n", self._env_name()]
                 )
+            )
 
-                history_file = os.path.join(info["prefix"], "conda-meta/history")
-                if not os.path.exists(history_file):
+            return os.path.join(info["prefix"], "conda-meta/history")
+
+        def _cache_valid(self, env_content):
+            try:
+                # Check if the env description has changed.
+                old_content = json.load(open(self._yaml_path()))
+                old_content_compare_key = json.dumps(old_content, sort_keys=True)
+                new_content_compare_key = json.dumps(env_content, sort_keys=True)
+                if old_content_compare_key != new_content_compare_key:
+                    print("detected change in environment definition.")
                     return False
 
-                history = open(history_file).readlines()
-                cmds = [line.strip() for line in history if line.startswith("# cmd: ")]
-                if len(cmds) != 1:
+                # Check if the setup commands have changed.
+                if self.setup_commands != json.load(
+                    open(self._setup_commands_snapshot_path())
```

Comment:
consider using
```
with open() as f:
     ....
```
So the file is closed once after loads

Reply:
The opened file will be closed when it goes out of scope, in this case once the sub-expression completes.

```py
import psutil

def main():
    proc = psutil.Process()
    print(proc.open_files())
    # []

    with open("/tmp/foo", "w") as f:
        print(proc.open_files())
        # [popenfile(path='/tmp/foo', fd=3, position=0, mode='w', flags=557057)]

    print(proc.open_files())
    # []

    open("/tmp/foo", "w")  # <-- no handle, so it closes at end of expression
    print(proc.open_files())
    # []

    f = open("/tmp/foo", "w")
    print(proc.open_files())
    # [popenfile(path='/tmp/foo', fd=3, position=0, mode='w', flags=557057)]

main()
```